src/Services/QueuedJobService.php

```diff
@@ -1533,6 +1579,271 @@ private function addJobHandlersToLogger(LoggerInterface $logger, QueuedJob $job,
         }
     }
 
+    /**
+     * Find any eligible jobs that qualify for an automated job retry and schedule their retry
+     *
+     * @throws ValidationException
+     */
+    protected function retryEligibleJobs(int $queueType): void
+    {
+        // Find any broken jobs that qualify for an automated job retry
+        $jobRetrySentinel = (int) static::config()->get('job_retry_sentinel');
+        $jobRetryLimit = (int) static::config()->get('job_retry_limit');
+        $jobRetryStatus = (array) static::config()->get('job_retry_status');
+
+        $jobStatusMap = [];
+        $jobRetrySentinel = max(0, $jobRetrySentinel);
+
+        // Job retries feature is disabled
+        if ($jobRetryLimit <= 0) {
+            return;
+        }
+
+        $validJobStatuses = QueuedJobDescriptor::singleton()->getJobStatusValues();
+
+        foreach ($jobRetryStatus as $retryStatus => $targetStatus) {
+            // Skip any disabled status conditions
+            if (!$targetStatus) {
+                continue;
+            }
+
+            // Skip any invalid status conditions
+            if (!in_array($targetStatus, $validJobStatuses)) {
+                continue;
+            }
+
+            $jobStatusMap[$retryStatus] = $targetStatus;
+        }
+
+        $jobStatusConditions = array_keys($jobStatusMap);
+
+        // We don't have any job status transformation configured
+        if (count($jobStatusConditions) === 0) {
+            return;
+        }
+
+        $now = DBDatetime::now();
+        $lastEditedSentinel = $now
+            // This is important to set otherwise it will break unit tests due to how the date mock works
+            ->setImmutable(true)
+            ->modify(sprintf('-%d minutes', $jobRetrySentinel))
+            ->Rfc2822();
+
+        $autoRetryJobs = QueuedJobDescriptor::get()
+            ->filter([
+                // Automatic retry is applied to only configured jobs statuses
+                'JobStatus' => $jobStatusConditions,
+                // Make sure to avoid retrying jobs that are being processed very recently to avoid edge cases
+                'LastEdited:LessThan' => $lastEditedSentinel,
+                // We only want to process jobs in the current queue
+                'JobType' => $queueType,
+            ])
+            // Oldest jobs first
+            ->sort('ID', 'ASC');
```

Comment:
```suggestion
            ->sort('Created', 'ASC');
```
While the result should be the same, this is more semantically clear. Since this column is used for sorting by default there is already an index for it so performance shouldn't be meaningfully impacted.

Reply:
Note that if we do keep the `job_retry_limit` configuration, always getting the oldest job first is probably not what we want (discussion about whether to keep that is in https://github.com/silverstripe/silverstripe-queuedjobs/pull/467#discussion_r2508366419).

Say we only allow 3 jobs (an intentionally low number for the sake of demonstration) to be retried per health check, but we have 20 jobs that need to be scheduled for retry. And we have a high `max_retry_attempts`, and the jobs keep failing so they keep needing to be retried.

If we always grab the oldest jobs first, the same 3 jobs will keep getting restarted every health check since they're always the oldest - we won't get to retry any of the other jobs until we've finally exausted the max retry attempts for those jobs.

If we do keep `job_retry_limit` we should probably be sorting this list by `LastEdited` (or `JobStarted` maybe, or even `CASE WHEN JobRestarted IS NOT NULL THEN JobRestarted ELSE JobStarted`), to get the jobs that were least recently retried.
Even that's not perfect, since different jobs will have different retry delays, and anything with a shorter retry delay may still end up sitting for longer than it should, while we reschedule other jobs with an earlier Last Edited date even if their delay is really long.

tl;dr I don't think we should keep `job_retry_limit` unless there's a really compelling reason for it (which can be discussed in [the comment thread for that config property](https://github.com/silverstripe/silverstripe-queuedjobs/pull/467#discussion_r2508366419) above)

Reply:
This shouldn't be a problem as long as the number of job retries per job is reasonably small number which it should be. Perhaps we should add a recommendation to the docs to not go overboard with the job retry limit? Probably higher than 5 is not recommended.